From 6e8d8feed2e5d6abc82f51336ecd386da6bee9df Mon Sep 17 00:00:00 2001
From: Jonas Gorski <jonas.gorski@bisdn.de>
Date: Mon, 13 May 2024 14:00:39 +0200
Subject: [PATCH 03/10] nexthop: add a identical helper function

Not all attributes of a nexthop are id attributes, e.g. the flags will
contain state (LINKDOWN, DEAD) of the attached link about which the
kernel will not send route updates.

Likewise, the weight may not exist when processing an ECMP IPv6 route
update which only contains a single nexthop.

Since rtnl_nexthop isn't a first class cache object, we cannot use
nl_object_identical(), so add a separate identical helper function which
compares only fixed attributes.

Upstream-Status: Backport [https://github.com/thom311/libnl/commit/8cf29d7b4a8f4dadcc68932edc23e05f5d0fee89]
[jonas.gorski: adapted for libnl 3.9]
Signed-off-by: Jonas Gorski <jonas.gorski@bisdn.de>
---
 include/netlink/route/nexthop.h |  3 +++
 lib/route/nexthop.c             | 17 +++++++++++++++++
 libnl-route-3.sym               |  5 +++++
 3 files changed, 25 insertions(+)

diff --git a/include/netlink/route/nexthop.h b/include/netlink/route/nexthop.h
index c4a2604aa8f8..1beb9fa9e679 100644
--- a/include/netlink/route/nexthop.h
+++ b/include/netlink/route/nexthop.h
@@ -30,6 +30,9 @@ extern int		rtnl_route_nh_compare(struct rtnl_nexthop *,
 					      struct rtnl_nexthop *,
 					      uint32_t, int);
 
+extern int		rtnl_route_nh_identical(struct rtnl_nexthop *,
+						struct rtnl_nexthop *);
+
 extern void		rtnl_route_nh_dump(struct rtnl_nexthop *,
 					   struct nl_dump_params *);
 
diff --git a/lib/route/nexthop.c b/lib/route/nexthop.c
index 962f2bab59c8..7e0df6136b04 100644
--- a/lib/route/nexthop.c
+++ b/lib/route/nexthop.c
@@ -136,6 +136,23 @@ int rtnl_route_nh_compare(struct rtnl_nexthop *a, struct rtnl_nexthop *b,
 	return diff;
 }
 
+/**
+ * Check if the fixed attributes of two nexthops are identical, and may
+ * only differ in flags or weight.
+ *
+ * @arg a		a nexthop
+ * @arg b		another nexthop
+ *
+ * @return true if both nexthop have equal attributes, otherwise false.
+ */
+int rtnl_route_nh_identical(struct rtnl_nexthop *a, struct rtnl_nexthop *b)
+{
+	return !rtnl_route_nh_compare(a, b,
+				      NH_ATTR_IFINDEX | NH_ATTR_REALMS |
+				      NH_ATTR_GATEWAY | NH_ATTR_NEWDST |
+				      NH_ATTR_VIA | NH_ATTR_ENCAP, 0);
+}
+
 static void nh_dump_line(struct rtnl_nexthop *nh, struct nl_dump_params *dp)
 {
 	struct nl_cache *link_cache;
diff --git a/libnl-route-3.sym b/libnl-route-3.sym
index fa7af4551495..a97ff3f74475 100644
--- a/libnl-route-3.sym
+++ b/libnl-route-3.sym
@@ -1314,3 +1314,8 @@ global:
 	rtnl_link_bond_set_min_links;
 	rtnl_link_can_get_device_stats;
 } libnl_3_8;
+
+libnl_3_10 {
+global:
+	rtnl_route_nh_identical;
+} libnl_3_9;
-- 
2.45.1

